src/state/routing/utils.ts

```diff
@@ -37,88 +55,113 @@ export function computeRoutes(
+      const isOnlyV3 = isVersionedRoute<V3PoolInRoute>(PoolType.V3Pool, route)
 
       return {
-        routev3:
-          routeProtocol === Protocol.V3
-            ? new V3Route(route.map(genericPoolPairParser) as Pool[], currencyIn, currencyOut)
-            : null,
-        routev2:
-          routeProtocol === Protocol.V2
-            ? new V2Route(route.map(genericPoolPairParser) as Pair[], currencyIn, currencyOut)
-            : null,
+        routev3: isOnlyV3 ? new V3Route(route.map(parsePool), parsedCurrencyIn, parsedCurrencyOut) : null,
+        routev2: isOnlyV2 ? new V2Route(route.map(parsePair), parsedCurrencyIn, parsedCurrencyOut) : null,
         mixedRoute:
-          routeProtocol === Protocol.MIXED
-            ? new MixedRouteSDK(route.map(genericPoolPairParser), currencyIn, currencyOut)
+          !isOnlyV3 && !isOnlyV2
+            ? new MixedRouteSDK(route.map(parsePoolOrPair), parsedCurrencyIn, parsedCurrencyOut)
             : null,
-        inputAmount: CurrencyAmount.fromRawAmount(currencyIn, rawAmountIn),
-        outputAmount: CurrencyAmount.fromRawAmount(currencyOut, rawAmountOut),
+        inputAmount: CurrencyAmount.fromRawAmount(parsedCurrencyIn, rawAmountIn),
+        outputAmount: CurrencyAmount.fromRawAmount(parsedCurrencyOut, rawAmountOut),
       }
     })
   } catch (e) {
-    // `Route` constructor may throw if inputs/outputs are temporarily out of sync
-    // (RTK-Query always returns the latest data which may not be the right inputs/outputs)
-    // This is not fatal and will fix itself in future render cycles
-    console.error(e)
-    return undefined
+    console.error('computeRoutes error', e)
+    return
   }
 }
 
-export function transformRoutesToTrade<TTradeType extends TradeType>(
-  route: ReturnType<typeof computeRoutes>,
-  tradeType: TTradeType,
-  blockNumber?: string | null,
-  gasUseEstimateUSD?: CurrencyAmount<Token> | null
-): InterfaceTrade<Currency, Currency, TTradeType> {
-  return new InterfaceTrade({
+const parsePoolOrPair = (pool: V3PoolInRoute | V2PoolInRoute): Pool | Pair => {
+  return pool.type === PoolType.V3Pool ? parsePool(pool) : parsePair(pool)
+}
+
+function isVersionedRoute<T extends V2PoolInRoute | V3PoolInRoute>(
+  type: T['type'],
+  route: (V3PoolInRoute | V2PoolInRoute)[]
+): route is T[] {
+  return route.every((pool) => pool.type === type)
+}
+
+export function transformRoutesToTrade(args: GetQuoteArgs, data: GetQuoteResult): TradeResult {
+  const { tokenInAddress, tokenOutAddress, tradeType } = args
+  const tokenInIsNative = Object.values(SwapRouterNativeAssets).includes(tokenInAddress as SwapRouterNativeAssets)
+  const tokenOutIsNative = Object.values(SwapRouterNativeAssets).includes(tokenOutAddress as SwapRouterNativeAssets)
+  const { gasUseEstimateUSD, blockNumber } = data
+  const routes = computeRoutes(tokenInIsNative, tokenOutIsNative, data.route)
+
+  const trade = new InterfaceTrade({
```

Comment:
Why is this type filtering necessary? Couldn't you just pass in the whole thing if you'd named them routev2, etc from computeRoutes?

Reply:
we can take a look at this logic more closely and move to sdk! i'd rather not make any changes for now, because this code is consistent with both widgets and mobile right now